README.md
[![travis](https://travis-ci.org/acburdine/gh-contrib-list.svg?branch=master)](https://travis-ci.org/acburdine/gh-contrib-list)
[![npm](https://img.shields.io/npm/v/gh-contrib-list.svg?maxAge=2592000)](https://npmjs.com/package/gh-contrib-list)

# gh-contrib-list

A tool to generate a list of contributors from a specific range of commits

## Usage:

```
npm install gh-contrib-list
```

In code:

```javascript
var ghContribList = require('gh-contrib-list');

ghContribList(options);
```

Returns a Promise with an array of contributors. Each contributor has the following properties:

- `id` - Github username of the contributor
- `name` - Full name of the contributor
- `commitCount` - Number of commits the person has contributed

### Options

- `user` - The user or organization (**required**)
- `repo` - The repository name (**required**)
- `commit` - A commit SHA to run the query from (commits including and after this one will be used) (**required**)
- `oauthKey` - If one is provided it will be used when making Github API requests
- `to` - If another commit SHA is provided here the commits used will be between the `commit` option and this one.
- `removeGreenkeeper` - If this is set to true, any commits by Greenkeeper will be automatically removed.
- `retry` - If this is set to true, the request will be retried in the event that the Github API returns a 202 status (retry momentarily).

---

## Credit:

This script is heavily based on [top-gh-contribs](https://github.com/novaugust/top-gh-contribs).


CHANGELOG.md
# 0.1.1

- [BUGFIX] last page of commits will now be added if commit is not found in list

# 0.1.0

- Initial version of script


test/index_spec.js
/* jshint expr:true */
var expect = require('chai').expect,
    nock = require('nock'),
    rewire = require('rewire'),

    contribList = rewire('../index.js');

describe('gh-contrib-list', function () {
    it('retry delay should be an exponential backoff in milliseconds', function () {
        var retryDelay = contribList.__get__('retryDelay');

        expect(retryDelay(0)).to.be.above(1000).and.below(2000);
        expect(retryDelay(1)).to.be.above(2000).and.below(3000);
        expect(retryDelay(2)).to.be.above(4000).and.below(5000);
        expect(retryDelay(3)).to.be.above(8000).and.below(9000);
        expect(retryDelay(4)).to.be.above(16000).and.below(17000);
    });

    describe('getPagination', function () {
        var getPagination = contribList.__get__('getPagination');

        afterEach(function () {
            nock.cleanAll();
        });

        it('should follow pagination links', function (done) {
            nock('https://api.github.com')
                .get('/repos/x/y/commits')
                .reply(
                    200,
                    {
                        id: 'acburdine',
                        name: 'Austin Burdine',
                        commitCount: 67
                    },
                    {
                        link: '<https://api.github.com/repositories/1234567/commits?page=2&per_page=100>; rel="next"'
                    }
                )
                .get('/repositories/1234567/commits?page=2&per_page=100')
                .reply(
                    200,
                    {
                        id: 'anotherUser',
                        name: 'Github User',
                        commitCount: 3
                    }
                );

            var opts = {
                url: 'http://api.github.com/repos/x/y/commits',
                userAgent: 'x'
            };

            getPagination(opts).then(function (results) {
                expect(results.length).to.equal(2);
                expect(results[0].id).to.equal('acburdine');
                expect(results[1].id).to.equal('anotherUser');

                done();
            }).catch(done);
        });
    });

    describe('getContribList', function () {
        var fixture = require(__dirname + '/fixtures.json');

        it('should return all unique contributors', function () {
            var result = contribList.getContribList(fixture);
            expect(result.length).to.equal(4);
        });

        it('should not return any greenkeeper commits if greenkeeper is disabled', function () {
            var result = contribList.getContribList(fixture, true);
            expect(result.length).to.equal(3);
        });

        it('should return the correct details for contributors (excluding merge commits)', function () {
            var result = contribList.getContribList(fixture, true);

            expect(result[0].commitCount).to.equal(7);
            expect(result[0].id).to.equal('kirrg001');
            expect(result[0].name).to.equal('Katharina Irrgang');
        });

        it('should return contributors in correct order', function () {
            var result = contribList.getContribList(fixture, true);
            expect(result[0].id).to.equal('kirrg001');
            expect(result[1].id).to.equal('kevinansfield');
            expect(result[2].id).to.equal('acburdine');
        });
    });

    describe('requestPromise', function () {
        var requestPromise = contribList.__get__('requestPromise');

        afterEach(function () {
            nock.cleanAll();
        });

        it('should reject on a status code of >= 500', function (done) {
            nock('http://example.com')
                .get('/')
                .reply(500);

            requestPromise({url: 'http://example.com/'}).then(function () {
                done(new Error('expected requestPromise to reject but it did not'));
            }).catch(function (err) {
                expect(err).to.be.instanceof(Error);

                done();
            });
        });

        it('should reject on a status code of >= 400', function (done) {
            nock('http://example.com')
                .get('/')
                .reply(404);

            requestPromise({url: 'http://example.com/'}).then(function () {
                done(new Error('expected requestPromise to reject but it did not'));
            }).catch(function (err) {
                expect(err).to.be.an.instanceof(Error);

                done();
            });
        });

        it('should return the response body on a status code of 200', function (done) {
            nock('http://example.com')
                .get('/')
                .reply(200, {some: 'thing'});

            requestPromise({url: 'http://example.com/'}).then(function (response) {
                expect(response).to.exist;
                expect(response[0].some).to.exist;
                expect(response[0].some).to.equal('thing');

                done();
            }).catch(done);
        });

        it('should return the response body and pagination url if available on a status code of 200', function (done) {
            nock('http://example.com')
                .get('/')
                .reply(200, {some: 'thing'}, {
                    link: '<https://api.github.com/repositories/8231436/commits?since=2015-05-08>; rel="next"'
                });

            requestPromise({url: 'http://example.com/'}).then(function (response) {
                expect(response).to.exist;
                expect(response[0].some).to.exist;
                expect(response[0].some).to.equal('thing');
                expect(response[1]).to.equal('https://api.github.com/repositories/8231436/commits?since=2015-05-08');

                done();
            }).catch(done);
        });

        it('should reject if status code is 202 and retry is not enabled', function (done) {
            nock('http://example.com')
                .get('/')
                .reply(202);

            requestPromise({url: 'http://example.com/'}).then(function () {
                done(new Error('expected requestPromise to reject but it did not'));
            }).catch(function (err) {
                expect(err).to.be.an.instanceof(Error);

                done();
            });
        });

        it('should retry if status code is 202 and retry is enabled', function (done) {
            this.timeout(5000);

            nock('http://example.com')
                .get('/')
                .reply(202)
                .get('/')
                .reply(200, {r: 'retry worked'});

            requestPromise({url: 'http://example.com/', retry: true}).then(function (response) {
                expect(response).to.exist;
                expect(response[0].r).to.exist;
                expect(response[0].r).to.equal('retry worked');

                done();
            }).catch(done);
        });
    });
});


index.js
'use strict';

var _ = require('lodash'),
    request = require('request'),
    Promise = require('bluebird');

function retryDelay(count) {
    return Math.floor((Math.pow(2, count) + Math.random()) * 1000);
}

/*
 * @param {Object} options
 * @param {string} url - the url to request
 * @param {string} [userAgent]
 * @param {string} [oauthKey] - a GitHub oauth key with access to the repository being queried
 * @param {boolean} [retry] - retry on status code 202
 * @param {number} [retryCount]
 */
function requestPromise(options) {
    options = options || {};
    var headers = {'User-Agent': options.userAgent || 'request'};

    if (options.oauthKey) {
        headers.Authorization = 'token ' + options.oauthKey;
    }

    return new Promise(function (resolve, reject) {
        request({
            url: options.url,
            json: true,
            headers: headers
        }, function (error, response, body) {
            if (error) {
                return reject(error);
            }

            // Check response headers for pagination links
            var links = response.headers.link,
                nextPageUrl = '';

            if (links && _.includes(links, 'next')) {
                nextPageUrl = links.substring(1, links.indexOf('>; rel="next'));
            }

            if (error) {
                return reject(error);
            }

            function decorateError(error) {
                if (!error) {
                    throw new Error('error is required.');
                }

                error.url = options.url;
                error.http_status = response.statusCode;
                error.ratelimit_limit = response.headers['x-ratelimit-limit'];
                error.ratelimit_remaining = response.headers['x-ratelimit-remaining'];
                error.ratelimit_reset = parseInt(response.headers['x-ratelimit-reset'], 10);

                return error;
            }

            if (response.statusCode >= 500) {
                return reject(decorateError(new Error('Server error on url ' + options.url)));
            }
            if (response.statusCode >= 400) {
                return reject(decorateError(new Error('Client error on url ' + options.url)));
            }
            if (response.statusCode === 202) {
                if (!options.retry || options.retryCount > 4) {
                    return reject(decorateError(new Error('API returned status 202. Try again in a few moments.')));
                }

                var retryCount = parseInt(options.retryCount, 10) || 0,
                    retryPromise = Promise.delay(retryDelay(retryCount)).then(function () {
                    return requestPromise({
                        url: options.url,
                        userAgent: options.userAgent || 'request',
                        oauthKey: options.oauthKey,
                        retry: true,
                        retryCount: retryCount + 1
                    });
                });

                return resolve(retryPromise);
            }
            return resolve([body, nextPageUrl]);
        });
    });
}

function getPagination(_commits, _options) {
    var commits, options;

    if (!_options) {
        options = _commits;
        commits = [];
    } else {
        commits = _commits;
        options = _options;
    }

    return requestPromise(options).then(function (results) {
        var indexOfLastSha = _.findIndex(results[0], ['sha', options.commit]);

        if (indexOfLastSha < 0) {
            commits = commits.concat(results[0]);

            if (results[1]) {
                options.url = results[1];
                return getPagination(commits, options);
            }

            return commits;
        }

        return commits.concat(_.slice(results[0], 0, indexOfLastSha + 1));
    });
}

function getContribList(commits, removeGreenkeeper) {
    commits = _(commits).filter(function (c) {
        var isValidCommit = (c.parents.length === 1 && c.author);

        if (removeGreenkeeper && isValidCommit) {
            return c.author.login !== 'greenkeeperio-bot';
        }

        return isValidCommit;
    })

    .map(function (c) {
        return {
            id: c.author.login,
            name: c.commit.author.name
        };
    })

    .reduce(function (contributors, commit) {
        var index = _.findIndex(contributors, {id: commit.id});

        if (index > -1) {
            contributors[index].commitCount += 1;
        } else {
            contributors.push({
                id: commit.id,
                name: commit.name,
                commitCount: 1
            });
        }
        return contributors;
    }, []);

    return _.orderBy(commits, ['commitCount'], ['desc']);
}

function main(options) {
    options = options || {};
    var user = options.user,
        repo = options.repo,
        commit = options.commit,
        oauthKey = options.oauthKey,
        removeGreenkeeper = options.removeGreenkeeper,
        queryParams = '?page=1&per_page=100';

    if (options.to) {
        queryParams += '&sha=' + options.to;
    }

    if (!(user && repo && commit)) {
        throw new Error('Must specify a github user, repo, and commit SHA');
    }

    var repoApiUrl = ['https://api.github.com/repos', user, repo, 'commits'].join('/'),
        paginationPromise = getPagination({
            url: repoApiUrl + queryParams,
            userAgent: user,
            commit: commit,
            oauthKey: oauthKey,
            retry: options.retry
        });

    return Promise.join(paginationPromise, removeGreenkeeper, getContribList);
}

main.getContribList = getContribList;
module.exports = main;
